pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
+        if 'y' in self.sources:
+            source.append('targets')
+        return source
+
+    def _get_canonical_spaces(self):
+        """
+        Returns a list. The first element will be the space of `X` if present,
+        or of `topo_view` otherwise. The second element will be the space of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        space = []
+        try:
+            space.append(self.spaces['X'])
+        except KeyError:
+            space.append(self.spaces['topo_view'])
+        try:
+            space.append(self.spaces['y'])
+        except KeyError:
+            pass
+        return space
+
+    def _get_canonical_data(self):
+        """
+        Returns a list. The first element will be the data of `X` if present,
+        or of `topo_view` otherwise. The second element will be the data of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
+        data = []
+        try:
+            data.append(self.data['X'])
+        except KeyError:
+            data.append(self.data['topo_view'])
+        try:
+            data.append(self.data['y'])
+        except KeyError:
+            pass
+        return data
 
-class HDF5DatasetIterator(FiniteDatasetIterator):
+    #@wraps(Dataset.get_data_specs, assigned=(), updated=())
+    # not in dataset, but required by iteration.755
+    def get_data_specs(self, source_or_alias=None):
+        """
+            Returns a tuple `(space, source)` related to the provided
+            source_or_alias key, if any, or to the canonical source otherwise.
+        """
+        if source_or_alias is None:
+            source_or_alias = self.aliases
 
-    """
-    Dataset iterator for HDF5 datasets.
-
-    FiniteDatasetIterator expects a design matrix to be available, but this
-    will not always be the case when using HDF5 datasets with topological
-    views.
-
-    Parameters
-    ----------
-    dataset : Dataset
-        Dataset over which to iterate.
-    subset_iterator : object
-        Iterator that returns slices of the dataset.
-    data_specs : tuple, optional
-        A (space, source) tuple.
-    return_tuple : bool, optional (default False)
-        Whether to return a tuple even if only one source is used.
-    convert : list, optional
-        A list of callables (in the same order as the sources in
-        data_specs) that will be applied to each slice of the dataset.
-    """
+        if isinstance(source_or_alias, tuple):
+            space = tuple([self.spaces[s] for s in source_or_alias])
+            space = CompositeSpace(space)
+        else:
+            space = self.spaces[source_or_alias]
+        return (space, source_or_alias)
 
-    def next(self):
+    #@wraps(Dataset.get, assigned=(), updated=())
+    # missing in dataset!
+    def get(self, sources, indexes):
```

Comment:
This seems like a but in the Dataset interface: _get_ is not part of the interface and I couldn't find it in iterator.py either.